sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,522 @@
+
+from ._online_lda import (mean_change, _dirichlet_expectation_1d,
+                         _dirichlet_expectation_2d)
+
+
+def _dirichlet_expectation(alpha):
+    """
+    For a vector theta ~ Dir(alpha), computes E[log(theta)] given alpha.
+    """
+    if len(alpha.shape) == 1:
+        ret = _dirichlet_expectation_1d(alpha)
+    else:
+        ret = _dirichlet_expectation_2d(alpha)
+    return ret
+
+
+def _update_gamma(X, expElogbeta, alpha, rng, max_iters,
+                  mean_change_tol, cal_delta):
+    """
+    E-step: update latent variable gamma
+    """
+
+    n_docs, n_vocabs = X.shape
+    n_topics = expElogbeta.shape[0]
+
+    # gamma is non-normailzed topic distribution
+    gamma = rng.gamma(100., 1. / 100., (n_docs, n_topics))
+    expElogtheta = np.exp(_dirichlet_expectation(gamma))
+    # diff on component (only calculate it when keep_comp_change is True)
+    delta_component = np.zeros(expElogbeta.shape) if cal_delta else None
+
+    X_data = X.data
+    X_indices = X.indices
+    X_indptr = X.indptr
+
+    for d in xrange(n_docs):
+        ids = X_indices[X_indptr[d]:X_indptr[d + 1]]
+        cnts = X_data[X_indptr[d]:X_indptr[d + 1]]
+        gammad = gamma[d, :]
+        expElogthetad = expElogtheta[d, :]
+        expElogbetad = expElogbeta[:, ids]
+        # The optimal phi_{dwk} is proportional to
+        # expElogthetad_k * expElogbetad_w. phinorm is the normalizer.
+        phinorm = np.dot(expElogthetad, expElogbetad) + 1e-100
+
+        # Iterate between gamma and phi until convergence
+        for it in xrange(0, max_iters):
+            lastgamma = gammad
+            # We represent phi implicitly to save memory and time.
+            # Substituting the value of the optimal phi back into
+            # the update for gamma gives this update. Cf. Lee&Seung 2001.
+            gammad = alpha + expElogthetad * \
+                np.dot(cnts / phinorm, expElogbetad.T)
+            expElogthetad = np.exp(_dirichlet_expectation(gammad))
+            phinorm = np.dot(expElogthetad, expElogbetad) + 1e-100
+
+            meanchange = mean_change(lastgamma, gammad)
+            if (meanchange < mean_change_tol):
+                break
+        gamma[d, :] = gammad
+        # Contribution of document d to the expected sufficient
+        # statistics for the M step.
+        if cal_delta:
+            delta_component[:, ids] += np.outer(expElogthetad, cnts / phinorm)
+
+    return (gamma, delta_component)
+
+
+class LatentDirichletAllocation(BaseEstimator, TransformerMixin):
+
+    """
+    Online Latent Dirichlet Allocation implementation with variational inference
+
+    References
+    ----------
+    [1] "Online Learning for Latent Dirichlet Allocation", Matthew D. Hoffman, 
+        David M. Blei, Francis Bach
+
+    [2] Matthew D. Hoffman's onlineldavb code. Link:
+        http://www.cs.princeton.edu/~mdhoffma/code/onlineldavb.tar
+
+
+    Parameters
+    ----------
+    n_topics: int, optional (default: 10)
```

Comment:
There needs to be a space before ``:`` for the numpy doc format (yes, everywhere, sorry).